overlord/snapstate/backend/link_test.go

```diff
@@ -566,3 +526,149 @@ type: os
 	c.Check(newCmdV6.Calls(), HasLen, 1)
 	c.Check(newCmdV7.Calls(), HasLen, 1)
 }
+
+func (s *linkCleanupSuite) testLinkCleanupFailedSnapdSnapOnCorePastWrappers(c *C, firstInstall bool) {
+	dirs.SetRootDir(c.MkDir())
+	defer dirs.SetRootDir("")
+
+	info, _ := mockSnapdSnapForLink(c)
+
+	err := os.MkdirAll(dirs.SnapServicesDir, 0755)
+	c.Assert(err, IsNil)
+	err = os.MkdirAll(dirs.SnapUserServicesDir, 0755)
+	c.Assert(err, IsNil)
+
+	// make snap mount dir non-writable, triggers error updating the current symlink
+	snapdSnapDir := filepath.Dir(info.MountDir())
+
+	if firstInstall {
+		err := os.Remove(filepath.Join(snapdSnapDir, "1234"))
+		c.Assert(err == nil || os.IsNotExist(err), Equals, true, Commentf("err: %v, err"))
+	} else {
+		err := os.Mkdir(filepath.Join(snapdSnapDir, "1234"), 0755)
+		c.Assert(err, IsNil)
+	}
+
+	// triggers permission denied error when symlink is manipulated
+	err = os.Chmod(snapdSnapDir, 0555)
+	c.Assert(err, IsNil)
+	defer os.Chmod(snapdSnapDir, 0755)
+
+	linkCtx := backend.LinkContext{
+		FirstInstall: firstInstall,
+	}
+	reboot, err := s.be.LinkSnap(info, mockDev, linkCtx, s.perfTimings)
+	c.Assert(err, ErrorMatches, fmt.Sprintf("symlink %s /.*/snapd/current: permission denied", info.Revision))
+	c.Assert(reboot, Equals, false)
+
+	checker := testutil.FilePresent
+	if firstInstall {
+		checker = testutil.FileAbsent
+	}
+
+	// system services
+	c.Check(filepath.Join(dirs.SnapServicesDir, "snapd.service"), checker)
+	c.Check(filepath.Join(dirs.SnapServicesDir, "snapd.socket"), checker)
+	c.Check(filepath.Join(dirs.SnapServicesDir, "snapd.snap-repair.timer"), checker)
+	// user services
+	c.Check(filepath.Join(dirs.SnapUserServicesDir, "snapd.session-agent.service"), checker)
+	c.Check(filepath.Join(dirs.SnapUserServicesDir, "snapd.session-agent.socket"), checker)
+	c.Check(filepath.Join(dirs.SnapServicesDir, "usr-lib-snapd.mount"), checker)
+}
+
+func (s *linkCleanupSuite) TestLinkCleanupFailedSnapdSnapOnCorePastWrappers(c *C) {
+	// test failure mode when snapd units were correctly written and
+	// corresponding services were started,
+	restore := release.MockOnClassic(false)
+	defer restore()
+
+	for _, isFirstInstall := range []bool{false, true} {
```

Comment:
this feels like it should just be two explicit tests invoking the helper, probably passing a context and not a bool

Reply:
This is split now